pkg/apis/configuration/v1alpha1/kongcustomentity_types.go

```diff
@@ -0,0 +1,76 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// KongCustomEntityList is a list of kong custom entities.
+type KongCustomEntityList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KongCustomEntity `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kongcustomentities,shortName=kce,categories=kong-ingress-controller
+
+// KongCustomEntity represents a custom entity in Kong.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Type is the type of this custom entity.
+	// Should be same as the `Name` of a KongCustomEntityDefinition.
+	Type string `json:"type"`
+	// Fields is the list of fields in the entity.
+	Fields apiextensionsv1.JSON `json:"fields"`
```

Comment:
I'd suggest putting these two fields into a separate struct `KongCustomEntitySpec` and also adding a separate `KongCustomEntityStatus` for keeping conditions.

If it comes to the conditions, I think we should populate `Programmed` similarly to other CRDs. Because of that, we should also require an ingress class to be specified for an entity to prevent multiple ingress controllers from writing into one object. I think the preferred way to define it would be by a `spec.ingressClass` field.

Reply:
done.